pkg/bucket/bucket_controller.go

```diff
@@ -0,0 +1,196 @@
+/* Copyright 2021 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucket
+
+import (
+	"context"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	kube "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	buckets "sigs.k8s.io/container-object-storage-interface-api/clientset"
+	bucketapi "sigs.k8s.io/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// BucketListener manages Bucket objects
+type BucketListener struct {
+	provisionerClient cosi.ProvisionerClient
+	provisionerName   string
+
+	kubeClient   kube.Interface
+	bucketClient buckets.Interface
+	kubeVersion  *utilversion.Version
+}
+
+// NewBucketListener returns a resource handler for Bucket objects
+func NewBucketListener(provisionerName string, client cosi.ProvisionerClient) *BucketListener {
+	bl := &BucketListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+	}
+
+	return bl
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
+// Return values
+//    nil - Bucket successfully provisioned
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (b *BucketListener) Add(ctx context.Context, bucket *v1alpha1.Bucket) error {
+	klog.V(3).InfoS("Add Bucket",
+		"name", bucket.Name,
+		"bucketclass", bucket.Spec.BucketClassName,
+	)
+
+	if !strings.EqualFold(bucket.Spec.Provisioner, b.provisionerName) {
```

Comment:
maybe this check could be move before the klog call above?

Reply:
yeah, you're right. Simple Oversight on my part

Reply:
It cannot be done because provisionerName can only be transitively obtained by getting the corresponding bucket.